src/main/java/com/redhat/lightblue/camel/lock/LightblueLockPolicy.java

```diff
@@ -56,57 +65,114 @@ public void beforeWrap(RouteContext routeContext, ProcessorDefinition<?> definit
+                Pair<T, String> pair = tryToLock(elements);
+
+                if (pair == null) {
+                    LOGGER.debug("Could not aquire a lock. Skipping processing.");
+                    return;
+                }
+
+                T lockedElement = pair.getLeft();
+                String lockedResourceId = pair.getRight();
+
+                exchange.getIn().setBody(lockedElement);
+
+                try {
+                    LOGGER.debug("Lock aquired, processing");
+                    exchange.getIn().setHeader(HEADER_LOCK_RESOURCE_ID, lockedResourceId);
+                    processor.process(exchange);
+                    exchange.getIn().removeHeader(HEADER_LOCK_RESOURCE_ID);
+                } finally {
+                    try{
+                        locking.release(lockedResourceId);
+                        if (LOGGER.isDebugEnabled())
+                            LOGGER.debug("Releasing lock on "+ java.net.URLDecoder.decode(lockedResourceId, "UTF-8"));
+                    }
+                    catch (Exception e) {
+                        if (exchange.isFailed()) {
+                            //Let the original exception bubble up, but log this one.
+                            LOGGER.error("Unexpected error while the route is already in a failed state.", e);
+                        } else {
+                            throw new LightblueLockingException(e);
                         }
                     }
                 }
-                else{
-                    LOGGER.debug("Unable to acquire a lock for: " + resourceId+" (it's taken)");
-                }
-
-                exchange.getIn().removeHeader(HEADER_LOCK_RESOURCE_ID);
             }
         };
 
     }
 
-    private Expression createRouteExpression(final RouteContext routeContext, final Expression expression){
-        ExpressionDefinition red = ExpressionNodeHelper.toExpressionDefinition(expression);
-        return red.createExpression(routeContext);
-    }
+    /**
+     * Tries to lock a random element recursively.
+     *
+     * @param elements
+     * @return Locked element and resourceId used to lock or null if locking was not possible.
+     */
+    private Pair<T, String> tryToLock(T[] elements) {
+        try {
+
+            int index;
+
+            if (elements.length > 1) {
+                index = randomGenerator.nextInt(elements.length);
+            } else {
+                index = 0;
+            }
+
+            T element = elements[index];
 
+            String resourceId = resourceIdExtractor.getResourceId(element);
+
+            if (LOGGER.isDebugEnabled())
+                LOGGER.debug("About to aquire lock on "+ java.net.URLDecoder.decode(resourceId, "UTF-8"));
+
+            if (locking.acquire(resourceId, ttl)) {
+                return new ImmutablePair<T, String>(element, resourceId);
+            } else if (elements.length > 1) {
+                T[] newElements = ArrayUtils.remove(elements, index);
```

Comment:
According to javadocs, this copies the input array. An unsuccessful locking attempt for an array of 100 will result 100 copies of the array with successively fewer elements.

A better approach here would be to add all the elements to an array list on method entry, and remove them during processing. Less memory intensive, easy to convert to loop.